presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -150,29 +150,41 @@ private static PlanNodeStatsEstimate createZeroStats(PlanNodeStatsEstimate stats
         StatisticRange add(StatisticRange leftRange, StatisticRange rightRange);
     }
 
+    @FunctionalInterface
+    private interface RowCountMergeStrategy
+    {
+        double merge(double left, double right);
+    }
+
     public static PlanNodeStatsEstimate addStatsAndSumDistinctValues(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right)
     {
-        return addStats(left, right, StatisticRange::addAndSumDistinctValues);
+        return addStats(left, right, StatisticRange::addAndSumDistinctValues, Double::sum);
     }
 
     public static PlanNodeStatsEstimate addStatsAndMaxDistinctValues(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right)
     {
-        return addStats(left, right, StatisticRange::addAndMaxDistinctValues);
+        return addStats(left, right, StatisticRange::addAndMaxDistinctValues, Double::sum);
     }
 
     public static PlanNodeStatsEstimate addStatsAndCollapseDistinctValues(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right)
     {
-        return addStats(left, right, StatisticRange::addAndCollapseDistinctValues);
+        return addStats(left, right, StatisticRange::addAndCollapseDistinctValues, Double::sum);
+    }
+
+    public static PlanNodeStatsEstimate addStatsAndIntersect(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right)
```

Comment:
This means that for the rowcount we'll use the min of the left and right sides.  I wonder if instead we should also filter based on percentage of overlapping ranges. 
e.g. if the column stats range for a column of one table are values 1-10, and the other table has a range of  5-20, then we would say maybe the new row count is the min of 1/2 the rows in the first table or 1/3 of the rows in the second table.  We could see problems for tables that have big ranges due to outliers (a common problem in all our stats rules, which assume a uniform distribution), so perhaps we should be more conservative that my suggestion so we don't do anything too terrible in that case, but aside from that I think it makes more sense.  @arhimondr what do you think?

Note: a difference between this and the union node is that the union node is really a UNION ALL node. we do unions as a union all + group by.

Reply:
@rschlussel You are right. In my logic, I was calculating worst-case number of rows (assuming random distribution). However, as you pointed out, the other estimators (filter node, inner join, etc.) assume uniform distribution.

I will update my estimation logic accordingly.

Reply:
@rschlussel At this point in code (https://github.com/prestodb/presto/pull/14143/files#diff-59fa6c2264eb72e0587e7e77ce155e59R187), row count is calculated for all columns together. So aggregated ColumnStats are unavailable. 

In the latest amend, I am evaluating stats for all columns and setting min distinct row count as row count for the intersect operator. Please have a look and let me know if that's correct.

Reply:
Using an estimated row count and then changing it at the end doesn't really work because it can make the null fraction and such be very far off.  I think instead of reusing the addStats method, you need to implement special logic for this case yourself.

You can use overlapPercentwith() from statisticRange in order to estimate the final rowcount (e.g. take the min overlap percent from the different columns and then multiply the rowcount by the overlap percent to get the final result). and then you can combine the column stats using intersect as you're doing already.

@arhimondr @ssaumitra what do you think?  The alternative is to go back to the original approach of setting the rowcount to the min of the two tables and not doing any range based filters.

Reply:
@rschlussel I did not understand side effect of updating row count completely. I would like to discuss it in more detail. Maybe I can open a conversation thread in Slack later day. Right now, I will just take your word for it.

I will give a shot to overlapPercentWith approach that you are talking about unless @arhimondr thinks differently.

Reply:
> This means that for the rowcount we'll use the min of the left and right sides. I wonder if instead we should also filter based on percentage of overlapping ranges.
e.g. if the column stats range for a column of one table are values 1-10, and the other table has a range of 5-20, then we would say maybe the new row count is the min of 1/2 the rows in the first table or 1/3 of the rows in the second table. We could see problems for tables that have big ranges due to outliers (a common problem in all our stats rules, which assume a uniform distribution), so perhaps we should be more conservative that my suggestion so we don't do anything too terrible in that case, but aside from that I think it makes more sense. @arhimondr what do you think?

I think it makes sense

Reply:
I have now updated row counting logic.